compiler/luci-interpreter/README.md

```diff
@@ -0,0 +1,137 @@
+
+### Inspecting intermediate state
+
+Interpreter provides interfaces to investigate internal state of interpreter during inference.
+
+This is done by "observer" mechanism:
+- `Interpreter` class has `attachObserver` method, which takes pointer to `ExecutionObserver` object
+- `ExecutionObserver` defines several callback methods user can override to inject custom code
+
+ExecutionObserver provides three callbacks:
+- `postTensorWrite` checks contents of output tensor after operation execution
+- `preOperatorExecute` notifies that interpreter is going to execute operation
+- `postOperatorExecute` notifies that interpreter finished operation execution
+
+See `luci-interpreter/include/luci_interpreter/Interpreter.h` for this interface details.
+
+Usage example:
+``` c++
+class CustomExecutionObserver: public luci_interpreter::ExecutionObserver
+{
+public:
+  void postTensorWrite(const luci::CircleNode *node, const Tensor *tensor) override
+  {
+    if (tensor->element_type() != loco::DataType::FLOAT32)
+      return;
+    for (int i = 0; i < tensor->shape().num_elements(); ++i)
+      std::cout << tensor->data<float>[i] << ", ";
+  }
+
+  // User observer can override only needed methods,
+  // others will inherit empty implementation from base observer.
+
+  // void preOperatorExecute(const luci::CircleNode *node);
+  // void postOperatorExecute(const luci::CircleNode *node);
+};
+
+luci_interpreter::Interpreter interpreter(module);
+CustomExecutionObserver observer;
+interpreter.attachObserver(&observer);
+
+// initialize input_data
+interpreter.writeInputTensor(input_node, input_data.data(), input_data.size());
+
+interpreter.interpret();
+```
+
+### Customizing inference
+
+#### Memory manager
+
+Interpreter provides handle to alter default memory management mechanisms.
+
+This is done by `MemoryManger` interface, see `luci-interpreter/include/luci_interpreter/MemoryManager.h` for implementation details.
+
+Header contains `IMemoryManager` abstract class which is responsible for allocation and dealocation of tensors memory.
+
+User can construct interpreter with one of predefined memory mmanagers or it's own custom memory manager.
+
+List of predefined memory managers:
+- `SimpleMemoryManager` This is simple wrapper around new/delete, default one.
+- `TestMemoryManager` Memorizing all allocated memory and releases it in Manager desctuctor, used in kernel unit tests.
+- `BuddyMemoryManager` Implements Buddy algorithm, uses external buffer for tensor data allocations, do not need new/delete.
+- `StaticMemoryManger` Uses precomputed memory allocation plan. Requires preparation with MemoryPlanner, but could improve memory consumption in restricted environments (like MCUs).
+
+Usage example:
+``` c++
+luci_interpreter::BuddyMemoryManager mm;
+
+luci_interpreter::Interpreter interpreter(module, &mm);
+
+// initialize input_data
+interpreter.writeInputTensor(input_node, input_data.data(), input_data.size());
+
+interpreter.interpret();
+...
+```
+
```

Comment:
```suggestion
BuddyMemoryManager usage example:
`TBD when it is merged`
```

Reply:
I hope it will be merged in near future =)

Reply:
done!